openedx/features/user_leads/tests/test_helpers.py

```diff
@@ -0,0 +1,114 @@
+import random
+
+from django.db import IntegrityError
+from django.test import TestCase
+from django.test.client import RequestFactory
+from django.urls import ResolverMatch
+
+from mock import patch
+from student.tests.factories import AnonymousUserFactory, UserFactory
+
+from ..constants import (
+    UTM_CAMPAIGN_KEY,
+    UTM_CONTENT_KEY,
+    UTM_MEDIUM_KEY,
+    UTM_PARAM_NAMES,
+    UTM_SOURCE_KEY,
+    UTM_TERM_KEY
+)
+from ..helpers import get_utm_params, save_user_utm
+from ..models import UserLeads
+
+
+class UserLeadsHelpersTest(TestCase):
+    """ Test cases for openedx user_leads feature helpers """
+
+    def setUp(self):
+        super(UserLeadsHelpersTest, self).setUp()
+
+        self.request_data = {
+            UTM_PARAM_NAMES[UTM_SOURCE_KEY]: u'test-source',
+            UTM_PARAM_NAMES[UTM_MEDIUM_KEY]: u'test-medium',
+            UTM_PARAM_NAMES[UTM_CONTENT_KEY]: u'test-content',
+            UTM_PARAM_NAMES[UTM_CAMPAIGN_KEY]: u'test-campaign',
+            UTM_PARAM_NAMES[UTM_TERM_KEY]: u'test-term',
+        }
+
+        self.user = UserFactory()
+        self.anonymous_user = AnonymousUserFactory()
+        self.url_path = '/dummy_path/'
+        self.url_name = 'dummy_name'
+
+    def test_get_utm_params_with_all_keys(self):
+        """
+        Test that a dict is returned with all the given utm params
+        extracted from request
+        """
+        request = RequestFactory().get(self.url_path, self.request_data)
+        returned_utm_params = get_utm_params(request)
+
+        self.assertDictEqual(returned_utm_params, self.request_data)
+
+    def test_get_utm_params_with_missing_key(self):
+        """
+        Test that a dict is returned without the missing utm param
+        """
+        utm_params = self.request_data.copy()
+        del utm_params[random.choice(list(utm_params.keys()))]
```

Comment:
Please verify, do we need to make a copy? `setUp` is called before each test, hence each test get complete data even if we delete some in any test

Reply:
You are right, thanks for the suggestion!